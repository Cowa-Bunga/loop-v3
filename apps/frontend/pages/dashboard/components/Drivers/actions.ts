import { DocumentData } from 'firebase/firestore'

const Actions = (_state, _setState) => ({
  modelDrivers: (documents: DocumentData[]): IDriver[] => {
    return documents.map((document) => ({
      id: document.id || document.NO_ID_FIELD,
      name: document.name,
      email: document.email,
      mobile_no: document.mobile_no,
      vehicle_type: document.vehicle_type,
      employee_code: document.employee_code,
      location: document.location,
      lunch: document.lunch,
      available: document.available,
      on_active_trip: document.on_active_trip,
      idle_since: document.idle_since,
      created_at: document.created_at
    }))
  },
  stringAvatar: (name: string) => {
    const fn = name.split(' ')[0][0]
    const ln = name.split(' ')[1] ? name.split(' ')[1][0] : ''

    return {
      sx: {
        bgcolor: _stringToColor(name)
      },
      children: `${fn}${ln}`
    }
  }
})

function _stringToColor(str: string) {
  let hash = 0
  let i

  /* eslint-disable no-bitwise */
  for (i = 0; i < str.length; i += 1) {
    hash = str.charCodeAt(i) + ((hash << 5) - hash)
  }

  let color = '#'

  for (i = 0; i < 3; i += 1) {
    const value = (hash >> (i * 8)) & 0xff
    color += `00${value.toString(16)}`.slice(-2)
  }
  /* eslint-enable no-bitwise */

  return color
}

export default Actions
