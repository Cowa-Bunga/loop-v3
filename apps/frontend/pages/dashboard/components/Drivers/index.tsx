import { Stack, Alert, Box, Drawer, TextField } from '@mui/material';
import ui from './style';

const Drivers = () => (
  <Drawer anchor="right" variant="permanent">
    <Box sx={ui.container}>
      <h4>Driver Activity</h4>
      <Box>
        <TextField
          fullWidth
          size="small"
          placeholder="Search driver name or code"
        />
        <Stack sx={ui.stack} spacing={2}>
          <Alert severity="warning">Available</Alert>
          <Alert severity="success">Busy</Alert>
        </Stack>
      </Box>
    </Box>
  </Drawer>
);

export default Drivers;
