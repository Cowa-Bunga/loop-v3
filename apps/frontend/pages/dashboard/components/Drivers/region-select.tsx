import { Alert, Box, Stack } from '@mui/material'
import ui from './style'

const RegionSelect = () => (
  <Box sx={ui.container}>
    <Stack sx={ui.stack} spacing={2}>
      <Alert severity="info">Select a region to view drivers</Alert>
    </Stack>
  </Box>
)

export default RegionSelect
