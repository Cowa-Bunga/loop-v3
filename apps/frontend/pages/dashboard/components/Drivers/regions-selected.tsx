import { useUserContext } from '@context/user'
import { collection, query, where } from '@util/lib/firebase'
import {
  useEffect,
  useFirestore,
  useFirestoreCollectionData,
  useMergeState
} from '@hooks'
import {
  Alert,
  Avatar,
  Box,
  CircularProgress,
  Input,
  List,
  ListItem,
  ListItemText,
  Stack
} from '@mui/material'
import ui from './style'
import Actions from './actions'

interface IProps {
  hubs: string[]
}

interface IState {
  availableDrivers: IDriver[]
  busyDrivers: IDriver[]
}

const RegionSelected = ({ hubs }: IProps) => {
  const { state: userContext } = useUserContext()
  const [state, setState] = useMergeState<IState>({
    availableDrivers: [],
    busyDrivers: []
  })
  const { modelDrivers, stringAvatar } = Actions(state, setState)

  const driversRef = collection(useFirestore(), 'drivers')
  const contains = hubs.map((id) => ({
    promisor_id: userContext.client.client_id,
    hub_id: id
  }))
  const constraints = where(
    'delivery_permissions',
    'array-contains-any',
    contains
  )
  const q = query(driversRef, constraints)
  const { status, data } = useFirestoreCollectionData(q)

  useEffect(() => {
    if (status == 'success') {
      const drivers = modelDrivers(data)
      setState({
        availableDrivers: drivers.filter((d: IDriver) => d.available),
        busyDrivers: drivers.filter((d: IDriver) => !d.available)
      })
    }
  }, [data, status])

  if (status !== 'success') {
    return (
      <Box sx={ui.container}>
        <CircularProgress color="secondary" />
      </Box>
    )
  }

  return (
    <Box sx={ui.container}>
      <Input
        sx={ui.searchDriver}
        fullWidth
        placeholder="Search driver name or code"
      />
      <Stack sx={ui.stack} spacing={2}>
        <Alert severity="success">Available</Alert>
        <List dense>
          {state.availableDrivers.map((driver) => (
            <ListItem key={driver.id}>
              <Avatar {...stringAvatar(driver.name)} />
              <div>&nbsp;</div>
              <ListItemText primary={driver.name} />
            </ListItem>
          ))}
        </List>
        <Alert severity="warning">Busy</Alert>
        <List dense>
          {state.busyDrivers.map((driver) => (
            <ListItem key={driver.id}>
              <Avatar {...stringAvatar(driver.name)} />
              <div>&nbsp;</div>
              <ListItemText primary={driver.name} />
            </ListItem>
          ))}
        </List>
      </Stack>
    </Box>
  )
}

export default RegionSelected
