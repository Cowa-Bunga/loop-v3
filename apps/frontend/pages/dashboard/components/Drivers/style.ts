export const ui = {
  container: {
    textAlign: 'center',
    width: '320px',
    minHeight: '100vh'
  },

  searchDriver: {
    color: 'white',
    mt: 4,
    p: 2,
    '& *': {
      color: 'white'
    }
  },

  stack: {
    mt: 4,
    width: '100%',
    p: 2
  }
}

export default ui
