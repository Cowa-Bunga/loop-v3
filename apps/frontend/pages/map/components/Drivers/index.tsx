import React from 'react'
import RegionSelect from './region-select'
import RegionSelected from './regions-selected'

interface IProps {
  hubs: string[]
}

const Drivers = ({ hubs }: IProps) => {
  if (hubs.length === 0) {
    return <RegionSelect />
  }

  return <RegionSelected hubs={hubs} />
}

export default Drivers
