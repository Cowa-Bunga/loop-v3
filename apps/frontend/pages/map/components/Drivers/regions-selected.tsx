import { useUserContext } from '@context/user'
import { collection, query, where } from 'firebase/firestore'
import { useFirestore, useFirestoreCollectionData } from 'reactfire'
import {
  Alert,
  Avatar,
  Box,
  CircularProgress,
  Input,
  List,
  ListItem,
  ListItemText,
  Stack
} from '@mui/material'
import ui from '@pages/map/components/Drivers/style'
import { useMergeState } from '@hooks'
import { useEffect } from 'react'
import Actions from './actions'

function stringToColor(string: string) {
  let hash = 0
  let i

  /* eslint-disable no-bitwise */
  for (i = 0; i < string.length; i += 1) {
    hash = string.charCodeAt(i) + ((hash << 5) - hash)
  }

  let color = '#'

  for (i = 0; i < 3; i += 1) {
    const value = (hash >> (i * 8)) & 0xff
    color += `00${value.toString(16)}`.slice(-2)
  }
  /* eslint-enable no-bitwise */

  return color
}

function stringAvatar(name: string) {
  const fn = name.split(' ')[0][0]
  const ln = name.split(' ')[1] ? name.split(' ')[1][0] : ''

  return {
    sx: {
      bgcolor: stringToColor(name)
    },
    children: `${fn}${ln}`
  }
}

interface IProps {
  hubs: string[]
}

interface IState {
  availableDrivers: IDriver[]
  busyDrivers: IDriver[]
}

const RegionSelected = ({ hubs }: IProps) => {
  const { state: userContext } = useUserContext()
  const [state, setState] = useMergeState<IState>({
    availableDrivers: [],
    busyDrivers: []
  })
  const { modelDrivers } = Actions(state, setState)

  const driversRef = collection(useFirestore(), 'drivers')
  const contains = hubs.map((id) => ({
    promisor_id: userContext.client.client_id,
    hub_id: id
  }))
  const constraints = where(
    'delivery_permissions',
    'array-contains-any',
    contains
  )
  const q = query(driversRef, constraints)
  const { status, data } = useFirestoreCollectionData(q)

  useEffect(() => {
    if (status == 'success') {
      const drivers = modelDrivers(data)
      setState({
        availableDrivers: drivers.filter((d: IDriver) => d.available),
        busyDrivers: drivers.filter((d: IDriver) => !d.available)
      })
    }
  }, [status])

  if (status !== 'success') {
    return (
      <Box sx={ui.container}>
        <CircularProgress color="secondary" />
      </Box>
    )
  }

  return (
    <Box sx={ui.container}>
      <Input
        sx={ui.searchDriver}
        fullWidth
        placeholder="Search driver name or code"
      />
      <Stack sx={ui.stack} spacing={2}>
        <Alert severity="success">Available</Alert>
        <List dense>
          {state.availableDrivers.map((driver) => (
            <ListItem key={driver.id}>
              <Avatar {...stringAvatar(driver.name)} />
              <div>&nbsp;</div>
              <ListItemText primary={driver.name} />
            </ListItem>
          ))}
        </List>
        <Alert severity="warning">Busy</Alert>
        <List dense>
          {state.busyDrivers.map((driver) => (
            <ListItem key={driver.id}>
              <Avatar {...stringAvatar(driver.name)} />
              <div>&nbsp;</div>
              <ListItemText primary={driver.name} />
            </ListItem>
          ))}
        </List>
      </Stack>
    </Box>
  )
}

export default RegionSelected
