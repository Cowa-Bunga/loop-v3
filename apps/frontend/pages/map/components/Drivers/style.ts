const ui = {
  container: {
    textAlign: 'center',
    width: '320px',
    height: '100vh'
  },

  searchDriver: {
    color: 'white',
    mt: 4,
    p: 2,
    '& *': {
      color: 'white'
    }
  },

  stack: {
    mt: 4,
    width: '100%',
    p: 2
  }
}

export default ui
