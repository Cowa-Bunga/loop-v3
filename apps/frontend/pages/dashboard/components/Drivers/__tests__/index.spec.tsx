import { renderWithProviders } from '@test/helpers'
import Drivers from '../'

jest.mock('./region-select', () => ({
  __esModule: true,
  default: () => <div>RegionSelect</div>
}))
jest.mock('./regions-selected', () => ({
  __esModule: true,
  default: () => <div>RegionsSelected</div>
}))
describe('Drivers', () => {
  it('should render without hubs', () => {
    const { baseElement } = renderWithProviders(<Drivers hubs={[]} />)
    expect(baseElement).toMatchSnapshot()
  })

  it('should render with hubs', () => {
    const { baseElement } = renderWithProviders(<Drivers hubs={['hub']} />)
    expect(baseElement).toMatchSnapshot()
  })
})
