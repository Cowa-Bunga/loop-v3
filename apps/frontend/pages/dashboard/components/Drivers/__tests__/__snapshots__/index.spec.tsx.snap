// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drivers should render with hubs 1`] = `
<body>
  <div>
    <div>
      RegionsSelected
    </div>
  </div>
</body>
`;

exports[`Drivers should render without hubs 1`] = `
<body>
  <div>
    <div>
      RegionSelect
    </div>
  </div>
</body>
`;
